packages/experiment-core/src/api/stream-api.ts

```diff
@@ -0,0 +1,194 @@
+      const randomReconnectionTimeout = Math.floor(
+        Math.random() * (MAX_CONN_MS_MAX - MAX_CONN_MS_MIN) + MAX_CONN_MS_MIN,
+      );
+      if (this.reconnectionTimeout) {
+        clearTimeout(this.reconnectionTimeout);
+      }
+      this.reconnectionTimeout = setTimeout(async () => {
+        if (es.readyState == this.eventSourceClass.OPEN) {
+          // The es is being checked, not this.eventSource. So it won't affect new connections.
+          this.close();
+          this.connect();
+        }
+      }, randomReconnectionTimeout);
+
+      // Set keep alive checks.
+      this.setKeepAliveExpiry();
+    };
+
+    // Timeout initial connection, ensures promise returns.
+    // Force close after timeout only if stream is still connecting.
+    // Error state should already handled by error handler.
+    if (this.initConnTimeout) {
+      clearTimeout(this.initConnTimeout);
+    }
+    this.initConnTimeout = setTimeout(() => {
+      es.readyState == this.eventSourceClass.CONNECTING && // The es is being checked, not this.eventSource. So it won't affect new connections.
+        this.error(DefaultStreamErrorEvents.TIMEOUT);
+    }, this.streamConnTimeoutMillis);
+  }
+
+  public close() {
+    if (this.eventSource) {
+      this.eventSource.close();
+      this.eventSource = undefined;
+    }
+
+    // Clear timers so program won't hang if want to terminate early.
+    if (this.initConnTimeout) {
+      clearTimeout(this.initConnTimeout);
+      this.initConnTimeout = undefined;
+    }
+    if (this.reconnectionTimeout) {
+      clearTimeout(this.reconnectionTimeout);
+      this.reconnectionTimeout = undefined;
+    }
+    if (this.keepAliveTimeout) {
+      clearTimeout(this.keepAliveTimeout);
+      this.keepAliveTimeout = undefined;
+    }
+  }
+
+  private async error(err: StreamErrorEvent) {
+    this.close();
+    if (this.onError) {
+      try {
+        await this.onError(err);
+        // eslint-disable-next-line no-empty
+      } catch {} // Don't care about errors after handoff.
+    }
+  }
+
+  private async handleNewMsg(response: StreamMessageEvent) {
+    if (!response.data) {
+      return;
+    }
+    this.setKeepAliveExpiry(); // Reset keep alive as there is data.
+    if (KEEP_ALIVE_INTERVAL > 0 && response.data == KEEP_ALIVE_DATA) {
+      // Data solely for keep alive. Don't pass on to client.
+      return;
+    }
+    if (this.onUpdate) {
+      try {
+        await this.onUpdate(response.data);
+        // eslint-disable-next-line no-empty
+      } catch {} // Don't care about errors after handoff.
+    }
+  }
+
+  private setKeepAliveExpiry() {
+    if (this.keepAliveTimeout) {
+      clearTimeout(this.keepAliveTimeout);
+      this.keepAliveTimeout = undefined;
+    }
+    if (KEEP_ALIVE_INTERVAL <= 0) {
+      return;
+    }
+    if (this.eventSource) {
+      this.keepAliveTimeout = setTimeout(() => {
+        this.error(DefaultStreamErrorEvents.KEEP_ALIVE_FAILURE);
+      }, KEEP_ALIVE_INTERVAL);
```

Comment:
Does the server define the keepalive interval? Is that set to 15 seconds? So this is 15 + 2 for buffer?

Reply:
Yes. server defines keepalive interval. I was thinking about making it configurable, but since it's server-defined, I guess there is no need for configuring it. I'll delete the checks.